### ISSUE TITLE
Feature request: Chatrooms with multiple characters

### ISSUE THREAD

**Issue by malfoyslastname** [agnaistic/agnai#115]
This is one of the biggest feature keeping certain enthusiasts on [the SillyLossy Tavern mod](https://github.com/SillyLossy/TavernAI).
This is also a character.ai feature.

UX-wise, what CAI does is that the bots just take turns speaking, but if you start typing in the input field (or click the pause/unpause) button, the bots stop talking until the input field is cleared.

It's unclear what the UX would be like in multi-user rooms. I'd suggest deferring support for multi-user, multi-bot rooms until requested.

Happy to collaborate.

# Example SFW outputs for motivation
<details>
  <summary>Click to reveal</summary>
  
  ![desu_eileen_room](https://user-images.githubusercontent.com/128472336/229193794-6283a725-37b5-48c6-892b-e2ab42242702.png)
  ![desu_esther_room](https://user-images.githubusercontent.com/128472336/229193800-e87071c6-8829-47a9-a038-2c6126e4a8df.png)
</details>

**Comment by malfoyslastname** [comment]
UX note: There should be a mode where users manually decide when a bot should send a message, and which bot. There should still be a mode where bots talk automatically. Some Tavern users report wanting a "smart" mode where which bot responds to the user's messages is determined based on who they're addressing. This is feasible, but needs more thought.

**Comment by acidbubbles** [comment]
I'm really looking forward to being able to do this, i.e. invite someone into an existing chat. Maybe to get things moving, there could be a way to:

1. Invite a character in a chat (the chat is still "owned" by one character)
2. Don't change anything to how chats work, but;
3. Where you added the Generate Image link, also add "Answer by CHAR" for each character

Once this exists, it'll be possible to make things like:

1. Ask who answers next every time, or;
2. Automatically guess based on things like whether the person was named, randomly, round robin, etc, or;
3. Ask an AI backend "who should answer next" :)

But all those could be separate improvements.

What do you think? Would that be something you'd merge? I might try myself at a "PoC PR" like I did for voice and events, see if that's interesting in practice.

**Comment by acidbubbles** [comment]
Actually I'm wondering if we should invite characters or have a room concept that is not associated with one character in particular...(refactor I guess)

**Comment by acidbubbles** [comment]
FYI I'm trying to implement this, at least enough to experiment with it.

![image](https://user-images.githubusercontent.com/34192666/233515707-92cac1d3-327d-4f25-b6b9-e50b444eaf9a.png)

**Comment by acidbubbles** [comment]
Oooooooooh I just realized what the main complexity was...

So, here's a branch with the work done (almost good enough for a PR) up to the invitation: https://github.com/acidbubbles/agn-ai/tree/multi-chars-chat

It's working but would need to also use websockets to publish the new character being added, see the characters in the list of participants, etc.

However......

What should the prompt look like?

I would say something like:

- The main character's scenario
- The participant character's personality and "talk like"

It also needs to:

- Change how messages work, since right now the code (generateResponseV2) takes from granted that the chat and the character are one entity, so changing that goes fairly deep

So @sceuick I'll pause this for now (though I might make an ugly implementation eventually just to play with it) and wait for some feedback on these:

1. The "scenario" should probably be separated from the character. For example, when I start a chat I imagine having a drop down with the "who", and another drop down with the  "scenario" (which is, again, related to the discussions in #116) that would probably include some world info, too. So I could decide to speak with John Doe in a café, with Jane Doe in a café, or with John Doe in a space adventure if I want. They are independent.
2. There should be a refactor to separate "chat" from "character". For example, if instead of "chat" we think of a "room" or "channel", you cannot take for granted that "send" means "reply with that character". Rather, "send" would be the user action, which is "only" sending a message (i.e. two users, no characters), AND have some logic that says "when a message is received, do X" where X might be responding automatically (single AI character), one of the options discussed previously (multiple AI characters) or nothing (no AI characters). This refactor should probably be done separately from the multi characters chat.

I don't think this is a huge change, but it's still big enough that I'd like to get your feedback before going further.

**Comment by malfoyslastname** [comment]
@acidbubbles Thank you so much for working on this, I'll try to take a look myself later, maybe we can also discuss on discord when convenient.

Prompt-wise, what I'd suggest doing is looking at how Silly does it because they do it very well. https://github.com/Cohee1207/SillyTavern

FWIW the way I'd wish "manual mode" to work UX wise is that you could choose in one click which bot to generates a response from. Something that would look like this:

![image](https://user-images.githubusercontent.com/128472336/233791557-15d83cde-7768-427e-9bca-d57e8b8f6591.png)


If the list of characters overflows from the viewport then it'd fall back to only showing pictures.

**Comment by acidbubbles** [comment]
That's exactly what I also had in mind :) From what I can see the hard part is to refactor messages to separate the chat from the character, right now send message and prompt character are tied up. Once separated, the rest will be fairly straightforward. I'm trying to avoid big changes right now though to avoid having huge merges with all the PRs :) I would really love to play with multiple characters though, it sounds amazing.

**Comment by acidbubbles** [comment]
... This is probably out of scope of this issue, but what I had in mind after manual mode was to ask the AI backend (maybe not the same as the one writing the text) who of the participants should write next based on the previous conversation text. I'd expect for example GPT-3.5 to do a fairly good job of just giving a name. I can imagine the AIs talking, and then turning towards you to ask you something for example. Or you could ask something to the characters without knowing who will answer, or address one specifically without having to then ask for an answer.

**Comment by R4y1a9n** [comment]
what would be really awesome is multiple characters loaded into chat room lets say its Bob Joe Sue and Pam.
they only talk when prompted. 
example: I see joe and bob by the door I wave as I go closer
(they are now in my chat) we have small chat 
I say I'll see you inside in a few and walk in the house I see Sue and Pam start talking to they
(now Bob Joe leave chat Sue and Pam join chat)
I ask sue have you met bob as I see him enter the room
(now bob joins chat with us) could even be a prompt like (Join <bob>) or (Leave <bob>) before starting to message
better yet drag and drop characters in and out of chat as needed!

**Comment by sceuick** [comment]
Resolved by #323
